### PR TITLE
Add ParameterName and PropertyName to TaskParameterEventArgs

### PR DESCRIPTION
**Pull request by ladipro**
Contributes to #9881

### Context

When `TaskParameterEventArgs` represents an output parameter of a task, it currently does not capture the name of the parameter. This information is generally useful as multiple output parameters can be assigned to the same item, and it's important for the BuildCheck infrastructure to provide a consistent object model to build analyzers.

Additionally, output parameters that are assigned to properties are currently logged as textual log messages `Output Property: {0}={1}` which miss the name of the parameter as well.

### Changes Made

Added two new properties on the `TaskParameterEventArgs` class, including the requisite serialization support. Updated logging to populate the properties with the name of the task parameter and the name of the build property, respectively, and updated BuildCheck infra to expose it to analyzers subscribed to task parameter events.

All task parameters are now logged as the structured `TaskParameterEventArgs` and the kind of parameter (input, output to property, output to item) can be determined by examining the args.

### Testing

- Tweaked existing unit test and updated the relevant BuildCheck test.
- Verified that when using the standard console logger, the resulting textual log is identical.
- Verified that binlog viewer without the corresponding changes displays the forward-compat warning about unknown data.

### Notes

- Complementary binlog viewer PR: https://github.com/KirillOsenkov/MSBuildStructuredLog/pull/780
- The change is behind a changewave because it is technically a breaking change for third party loggers.

### CODE REVIEW

**Review by JanKrivanek — APPROVED** [review]
Thank you!

File: src/Build.UnitTests/BuildEventArgsSerialization_Tests.cs [hunk 1]
```diff
@@ -650,12 +650,14 @@ public void RoundtripTaskParameterEventArgs()
                 new TaskItemData("ItemSpec1", null),
                 new TaskItemData("ItemSpec2", Enumerable.Range(1,3).ToDictionary(i => i.ToString(), i => i.ToString() + "value"))
             };
-            var args = new TaskParameterEventArgs(TaskParameterMessageKind.TaskOutput, "ItemName", items, true, DateTime.MinValue);
+            var args = new TaskParameterEventArgs(TaskParameterMessageKind.TaskOutput, "ParameterName", "PropertyName", "ItemName", items, true, DateTime.MinValue);
```
[the remark this answers is not in the document]
Comment (ladipro) [reply src/Build.UnitTests/BuildEventArgsSerialization_Tests.cs:653]: Makes sense, thank you.

**Review by KirillOsenkov — APPROVED** [review]

File: src/Build/BackEnd/TaskExecutionHost/TaskExecutionHost.cs [hunk 2]
```diff
@@ -1320,7 +1320,9 @@ private void EnsureParameterInitialized(TaskPropertyInfo parameter, Lookup looku
                         ItemGroupLoggingHelper.LogTaskParameter(
                             _taskLoggingContext,
                             TaskParameterMessageKind.TaskInput,
-                            parameter.Name,
+                            parameterName: parameter.Name,
+                            propertyName: null,
+                            itemType: parameter.Name,
```
[the remark this answers is not in the document]
Comment (ladipro) [reply src/Build/BackEnd/TaskExecutionHost/TaskExecutionHost.cs:1325]: Apologies, yes, I forgot about this. Passing the parameter name in `ItemType` makes the old viewer display correct data (at least for input parameters). So basically a back-compat hack. I'm on the fence - we can make this cleaner if we're fine breaking old viewers. We're breaking them for output parameters anyway.

[the remark this answers is not in the document]
Comment (ladipro) [reply src/Build/BackEnd/TaskExecutionHost/TaskExecutionHost.cs:1327]: Ok, I am going with the back compat plan. `TaskParameterEventArgs` will continue to have its `ItemType` used for different things to keep old loggers working and old binlog viewer show everything. Enlightened loggers will use the newly added props to disambiguate.

**Pull request merged**